### PR TITLE
Get recordings merged enhancements

### PR DESCRIPTION
**Pull request by jfederico**
Closed pull requests #2981 & #2987 in order to open this one with both merged

### CODE REVIEW

**Comment by ffdixon** [comment]
On one of our testing servers we're noticing that tomcat7 is throwing a lot of errors related to too many open files

    java.nio.file.FileSystemException: /var/bigbluebutton/unpublished: Too many open files
        at sun.nio.fs.UnixException.translateToIOException(UnixException.java:91)
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
        at sun.nio.fs.UnixFileSystemProvider.newDirectoryStream(UnixFileSystemProvider.java:426)
        at java.nio.file.Files.newDirectoryStream(Files.java:413)
        at org.bigbluebutton.api.RecordingService.getDirectories(RecordingService.java:228)
        at org.bigbluebutton.api.RecordingService.getPlaybackFormats(RecordingService.java:241)
        at org.bigbluebutton.api.RecordingService.getAllDirectories(RecordingService.java:365)
        at org.bigbluebutton.api.RecordingService.getAllDirectories(RecordingService.java:383)
        at org.bigbluebutton.api.RecordingService.getRecordings(RecordingService.java:72)
        at org.bigbluebutton.api.MeetingService.getRecordings(MeetingService.java:356)
        at org.bigbluebutton.api.MeetingService$getRecordings$0.call(Unknown Source)
        at org.bigbluebutton.web.controllers.ApiController$_closure19.doCall(ApiController.groovy:1728)
        at org.bigbluebutton.web.controllers.ApiController$_closure19.doCall(ApiController.groovy)
        at sun.reflect.GeneratedMethodAccessor272.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:90)
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:233)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1086)
        at groovy.lang.ExpandoMetaClass.invokeMethod(ExpandoMetaClass.java:1110)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:910)
        at groovy.lang.Closure.call(Closure.java:411)
        at groovy.lang.Closure.call(Closure.java:405)
        at org.bigbluebutton.web.controllers.ApiController.getRecordingsHandler(ApiController.groovy)
        at sun.reflect.GeneratedMethodAccessor271.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.codehaus.groovy.grails.web.servlet.mvc.MixedGrailsControllerHelper.invoke(MixedGrailsControllerHelper.java:154)
        at org.codehaus.groovy.grails.web.servlet.mvc.AbstractGrailsControllerHelper.handleAction(AbstractGrailsControllerHelper.java:357)
        at org.codehaus.groovy.grails.web.servlet.mvc.AbstractGrailsControllerHelper.executeAction(AbstractGrailsControllerHelper.java:234)
        at org.codehaus.groovy.grails.web.servlet.mvc.AbstractGrailsControllerHelper.handleURI(AbstractGrailsControllerHelper.java:200)
        at org.codehaus.groovy.grails.web.servlet.mvc.AbstractGrailsControllerHelper.handleURI(AbstractGrailsControllerHelper.java:121)
        at org.codehaus.groovy.grails.web.servlet.mvc.SimpleGrailsController.handleRequest(SimpleGrailsController.java:72)
        at org.springframework.web.servlet.mvc.SimpleControllerHandlerAdapter.handle(SimpleControllerHandlerAdapter.java:48)
        at org.codehaus.groovy.grails.web.servlet.GrailsDispatcherServlet.doDispatch(GrailsDispatcherServlet.java:355)
        at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:856)
        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:953)
        at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:844)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:620)
        at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:829)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:727)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
        at grails.plugin.cache.web.filter.PageFragmentCachingFilter.doFilter(PageFragmentCachingFilter.java:200)
        at grails.plugin.cache.web.filter.AbstractFilter.doFilter(AbstractFilter.java:63)
        at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:343)
        at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:260)


When I use lsof -p on tomcat process to see what files it has open, I see thousands of

	java    1799 tomcat7 4060r   DIR                9,3     4096 50331674 /home/bigbluebutton/bigbluebutton/unpublished
	java    1799 tomcat7 4061r   DIR                9,3     4096 50331674 /home/bigbluebutton/bigbluebutton/unpublished
	java    1799 tomcat7 4062r   DIR                9,3     4096 50594769 /home/bigbluebutton/bigbluebutton/unpublished/presentation
	java    1799 tomcat7 4063r   DIR                9,3     4096 50594769 /home/bigbluebutton/bigbluebutton/unpublished/presentation
	java    1799 tomcat7 4064r   DIR                9,3    61440 50331766 /home/bigbluebutton/bigbluebutton/published/presentation
	java    1799 tomcat7 4065r   DIR                9,3    61440 50331766 /home/bigbluebutton/bigbluebutton/published/presentation
	java    1799 tomcat7 4066r   DIR                9,3     4096 50331674 /home/bigbluebutton/bigbluebutton/unpublished
	java    1799 tomcat7 4067r   DIR                9,3     4096 50331674 /home/bigbluebutton/bigbluebutton/unpublished
	java    1799 tomcat7 4068r   DIR                9,3     4096 50594769 /home/bigbluebutton/bigbluebutton/unpublished/presentation
	java    1799 tomcat7 4069r   DIR                9,3     4096 50594769 /home/bigbluebutton/bigbluebutton/unpublished/presentation
	java    1799 tomcat7 4070r   DIR                9,3     4096 50331672 /home/bigbluebutton/bigbluebutton/published
	java    1799 tomcat7 4071r   DIR                9,3     4096 50331672 /home/bigbluebutton/bigbluebutton/published
	java    1799 tomcat7 4072r   DIR                9,3     4096 50331672 /home/bigbluebutton/bigbluebutton/published
	java    1799 tomcat7 4073r   DIR                9,3     4096 50331672 /home/bigbluebutton/bigbluebutton/published
	java    1799 tomcat7 4074r   DIR                9,3    61440 50331766 /home/bigbluebutton/bigbluebutton/published/presentation
	java    1799 tomcat7 4075r   DIR                9,3    61440 50331766 /home/bigbluebutton/bigbluebutton/published/presentation
	java    1799 tomcat7 4076r   DIR                9,3     4096 50331674 /home/bigbluebutton/bigbluebutton/unpublished
	java    1799 tomcat7 4077r   DIR                9,3     4096 50331674 /home/bigbluebutton/bigbluebutton/unpublished
	java    1799 tomcat7 4078r   DIR                9,3     4096 50594769 /home/bigbluebutton/bigbluebutton/unpublished/presentation
	java    1799 tomcat7 4079r   DIR                9,3     4096 50594769 /home/bigbluebutton/bigbluebutton/unpublished/presentation


In looking at

    http://stackoverflow.com/questions/26334421/java-nio-file-filesystemexception-proc-too-many-open-files

I'm wondering if the new code to iterate through recordings is closing file descriptors when it's looping?

File: bigbluebutton-web/src/java/org/bigbluebutton/api/RecordingService.java [hunk 1]
```diff
@@ -1,259 +1,405 @@
...
+
+    private List<File> getDirectories(String path) {
+        List<File> files = new ArrayList<File>();
+        try {
+            DirectoryStream<Path> stream = Files.newDirectoryStream(FileSystems.getDefault().getPath(path));
+            Iterator<Path> iter = stream.iterator();
+            while (iter.hasNext()) {
+                Path next = iter.next();
+                files.add(next.toFile());
+            }
```
Comment (fcecagno) [inline bigbluebutton-web/src/java/org/bigbluebutton/api/RecordingService.java]: Related to this comment: https://github.com/bigbluebutton/bigbluebutton/pull/3001#issuecomment-181380690

Found this: https://docs.oracle.com/javase/7/docs/api/java/nio/file/DirectoryStream.html
>  A DirectoryStream is opened upon creation and is closed by invoking the close method. Closing a directory stream releases any resources associated with the stream. Failure to close the stream may result in a resource leak. The try-with-resources statement provides a useful construct to ensure that the stream is closed:
>
>   Path dir = ...
>   try (DirectoryStream<Path> stream = Files.newDirectoryStream(dir)) {
>       for (Path entry: stream) {
>           ...
>       }
>   }
So we should revisit this code portion to release the resources properly.

**Comment by fcecagno** [comment]
@ffdixon You're right, I just added a comment where the issue is located: https://github.com/bigbluebutton/bigbluebutton/pull/3001/files#r52172277

**Comment by jfederico** [comment]
The stream was never closed. Done

https://github.com/jfederico/bigbluebutton/commit/c40f499b9462580a0b1091b6fb1c88d317a6102e

**Pull request merged**